prow/plugins/projectqueries/projectqueries.go

```diff
@@ -0,0 +1,208 @@
+func labelsInLabels(queryLabels []string, labels []github.Label) bool {
+	for _, queryLabel := range queryLabels {
+		match := false
+		for _, label := range labels {
+			if queryLabel == label.Name {
+				match = true
+				break
+			}
+		}
+		if match == false {
+			return false
+		}
+	}
+	return true
+}
+
+// These are the only actions that this plugin cares about
+func isPRChanged(pe github.PullRequestEvent) bool {
+	switch pe.Action {
+	case github.PullRequestActionOpened:
+		return true
+	case github.PullRequestActionClosed:
+		return true
+	case github.PullRequestActionReopened:
+		return true
+	case github.PullRequestActionLabeled:
+		return true
+	case github.PullRequestActionUnlabeled:
+		return true
+	default:
+		return false
+	}
+}
+
+func getMatchingProjectQueries(gc githubClient, projectQueries []plugins.ProjectQuery, pe github.PullRequestEvent) ([]plugins.ProjectQuery, error) {
+	matchedProjectQueries := make([]plugins.ProjectQuery, 0, 1)
+	// GetIssueLabels works for PRs as they are considered issues in the API
+	labels, err := gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
+	if err != nil {
+		return nil, err
+	}
+	for _, projectQuery := range projectQueries {
+		searchQuery := projectQuery.Query
+		if searchQuery.Org != "" && searchQuery.Org != pe.Repo.Owner.Login {
+			continue
+		}
+		if searchQuery.State != "" && searchQuery.State != pe.PullRequest.State {
+			continue
+		}
+		if len(searchQuery.Labels) != 0 && !labelsInLabels(searchQuery.Labels, labels) {
+			continue
+		}
+		matchedProjectQueries = append(matchedProjectQueries, projectQuery)
+	}
+	return matchedProjectQueries, nil
+}
+
+func addPRToProject(gc githubClient, projectQuery plugins.ProjectQuery, pe github.PullRequestEvent) error {
+	// Get the project id by name
+	projects, err := gc.GetOrgProjects(projectQuery.Org)
+	if err != nil {
+		return err
+	}
+	projectID := -1
+	for _, project := range projects {
+		if project.Name == projectQuery.Name {
+			projectID = project.ID
+			break
+		}
+	}
+	if projectID == -1 {
+		return fmt.Errorf("could not find project %s in org %s", projectQuery.Name, projectQuery.Org)
+	}
+
+	// Get the project columns by project id
+	projectColumns, err := gc.GetProjectColumns(projectID)
+	if err != nil {
+		return err
+	}
+	if len(projectColumns) == 0 {
+		return fmt.Errorf("no columns in project %s with id %d", projectQuery.Name, projectID)
+	}
+
+	// Default to using first column of project
+	columnID := projectColumns[0].ID
+	// If a column name is given try and find that column
+	if projectQuery.Column != "" {
+		for _, projectColumn := range projectColumns {
+			if projectQuery.Column == projectColumn.Name {
+				columnID = projectColumn.ID
+				break
+			}
+		}
+	}
+
+	// Create project card and add this PR
+	projectCard := github.ProjectCard{}
+	projectCard.ContentType = "PullRequest"
+	projectCard.ContentID = pe.PullRequest.ID
+	_, err = gc.CreateProjectCard(columnID, projectCard)
```

Comment:
What happens if there exists a project definition that is owned by this plugin, and the PR used to match the criteria but no longer does? Do we want to remove it?

Reply:
At the moment it just adds pull requests. Removing them or moving them along the columns has to be done manually.

If it could remove pull requests that would be nicer, and would even mean things could progress through columns automatically.

Removing them would mean either keeping track of automatically added pull requests, or removing any pull request that doesn't match the query. Removing anything that doesn't match could be annoying, as if for some reason a pull request was manually added that doesn't match the query it would be removed by the plugin.

Reply:
Can we declare that for Projects managed by this plugin, only automatic PR additions are allowed? What do we need for the UX here?

Reply:
@cblecker